provider/openstack/cinder.go

```diff
@@ -0,0 +1,443 @@
+		if !ok {
+			return nil, errors.Errorf("could not find volume: %s", volId)
+		}
+		volumes[volIdx] = cinderVol
+	}
+
+	return volumes, nil
+}
+
+// DestroyVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
+			errors[volIdx] = err
+		}
+	}
+	return errors
+}
+
+// ValidateVolumeParams implements storage.VolumeSource.
+func (s *openstackVolumeSource) ValidateVolumeParams(params storage.VolumeParams) error {
+	return nil
+}
+
+// AttachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) AttachVolumes(args []storage.VolumeAttachmentParams) ([]storage.VolumeAttachment, error) {
+
+	volAttachments := make([]storage.VolumeAttachment, 0, len(args))
+	for _, attachArg := range args {
+
+		// Check to see if the volume is already attached.
+		existingVolAttachments, err := s.openstackClient.ListVolumeAttachments(string(attachArg.InstanceId))
+		if err != nil {
+			return nil, err
+		}
+		if findAttachment(attachArg.VolumeId, existingVolAttachments) != nil {
+			continue
+		}
+
+		// A volume's status must be available before it can be
+		// attached.
+		const numAttempts = 10
+		const retryInterval = 5 * time.Second
+		if err := <-s.openstackClient.VolumeStatusNotifier(
+			attachArg.VolumeId,
+			"available",
+			numAttempts,
+			retryInterval,
+		); err != nil {
+			return nil, err
+		}
+
+		attachment, err := s.openstackClient.AttachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			autoAssignedMountPoint,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		volAttachments = append(volAttachments, attachment)
+	}
+
+	return volAttachments, nil
+}
+
+// DetachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DetachVolumes(args []storage.VolumeAttachmentParams) error {
+
+	for _, attachArg := range args {
+		// Check to see if the volume is already detached.
+		volAttachments, err := s.openstackClient.ListVolumeAttachments(string(attachArg.InstanceId))
+		if err != nil {
+			return err
+		}
+
+		if err := detachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			volAttachments,
+			s.openstackClient,
+		); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type OpenstackAdapter interface {
+	GetVolumesSimple() ([]storage.Volume, error)
+	DeleteVolume(string) error
+	CreateVolume(size uint64, tag names.VolumeTag) (storage.Volume, error)
+	AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error)
+	VolumeStatusNotifier(volId, status string, numAttempts int, waitDur time.Duration) <-chan error
+	DetachVolume(serverId, attachmentId string) error
+	ListVolumeAttachments(serverId string) ([]storage.VolumeAttachment, error)
+}
+
+func NewGooseAdapter(environConfig *config.Config) (OpenstackAdapter, error) {
```

Comment:
Either this should be newOpenstackAdapter, s/newOpenstackAdapter/NewGooseAdapter/g

Not sure this should be exported anyway...